src/browser/fetch/fetch.zig

```diff
@@ -0,0 +1,231 @@
+const HttpClient = @import("../../http/Client.zig");
+const Mime = @import("../mime.zig").Mime;
+
+const Headers = @import("Headers.zig");
+
+const RequestInput = @import("Request.zig").RequestInput;
+const RequestInit = @import("Request.zig").RequestInit;
+const Request = @import("Request.zig");
+const Response = @import("Response.zig");
+
+pub const Interfaces = .{
+    @import("Headers.zig"),
+    @import("Headers.zig").HeadersEntryIterable,
+    @import("Headers.zig").HeadersKeyIterable,
+    @import("Headers.zig").HeadersValueIterable,
+    @import("Request.zig"),
+    @import("Response.zig"),
+};
+
+pub const FetchContext = struct {
+    arena: std.mem.Allocator,
+    js_ctx: *Env.JsContext,
+    promise_resolver: v8.Persistent(v8.PromiseResolver),
+
+    method: Http.Method,
+    url: []const u8,
+    body: std.ArrayListUnmanaged(u8) = .empty,
+    headers: std.ArrayListUnmanaged([]const u8) = .empty,
+    status: u16 = 0,
+    mime: ?Mime = null,
+    transfer: ?*HttpClient.Transfer = null,
+
+    /// This effectively takes ownership of the FetchContext.
+    ///
+    /// We just return the underlying slices used for `headers`
+    /// and for `body` here to avoid an allocation.
+    pub fn toResponse(self: *const FetchContext) !Response {
+        var headers: Headers = .{};
+
+        // convert into Headers
+        for (self.headers.items) |hdr| {
+            var iter = std.mem.splitScalar(u8, hdr, ':');
+            const name = iter.next() orelse "";
+            const value = iter.next() orelse "";
+            try headers.append(name, value, self.arena);
+        }
+
+        return Response{
+            .status = self.status,
+            .headers = headers,
+            .mime = self.mime,
+            .body = self.body.items,
+            .url = self.url,
+        };
+    }
+
+    pub fn destructor(self: *FetchContext) void {
+        if (self.transfer) |_| {
+            const resolver = Env.PromiseResolver{
+                .js_context = self.js_ctx,
+                .resolver = self.promise_resolver.castToPromiseResolver(),
+            };
+
+            resolver.reject("TypeError") catch unreachable;
+            self.promise_resolver.deinit();
+        }
+    }
+};
+
+// https://developer.mozilla.org/en-US/docs/Web/API/Window/fetch
+pub fn fetch(input: RequestInput, options: ?RequestInit, page: *Page) !Env.Promise {
+    const arena = page.arena;
+
+    const req = try Request.constructor(input, options, page);
+
+    const resolver = Env.PromiseResolver{
+        .js_context = page.main_context,
+        .resolver = v8.PromiseResolver.init(page.main_context.v8_context),
+    };
+
+    var headers = try Http.Headers.init();
+
+    // Copy our headers into the HTTP headers.
+    var header_iter = req.headers.headers.iterator();
+    while (header_iter.next()) |entry| {
+        // This is fine because curl/headers copies it internally.
+        const combined = try std.fmt.allocPrintSentinel(
+            page.call_arena,
```

Comment:
I _think_ this might have to be `page.arena`, because the http client has a queue, so the headers might not be passed off to libcurl within this function call.